test/e2e/alert_relabel_config_test.go

```diff
@@ -56,124 +61,70 @@ func TestAlertRelabelConfigTechPreview(t *testing.T) {
+	// Try to create an invalid AlertRelabelConfig.
+	_, err = relabelConfigs.Create(ctx, arc, metav1.CreateOptions{})
+	if !apierrors.IsInvalid(err) {
+		t.Fatal(errors.Wrap(err, "invalid AlertRelabelConfig wasn't rejected."))
 	}
 
-	// Check that it is added to the secret.
-	err = framework.Poll(time.Second, 2*time.Minute, func() error {
-		s, err := secrets.Get(ctx, relabelSecretName, metav1.GetOptions{})
-		if err != nil {
-			return err
-		}
-
-		var configsFromSecret []relabel.Config
-		err = yaml.Unmarshal([]byte(s.Data[relabelSecretKey]), &configsFromSecret)
-		if err != nil {
-			t.Fatal(errors.Wrap(err, "failed to unmarshal AlertRelabelConfig from secret"))
-		}
-
-		err = relabelConfigsEqual(arc.Spec.Configs[0], configsFromSecret[0])
-		if err != nil {
-			t.Fatal(errors.Wrap(err, "AlertRelabelConfig from secret doesn't match resource"))
-		}
-
-		return nil
-	})
-
+	// Create a valid AlertRelabelConfig.
+	arc.Spec.Configs[0].Action = "Replace"
+	_, err = relabelConfigs.Create(ctx, arc, metav1.CreateOptions{})
 	if err != nil {
-		t.Fatal(errors.Wrap(err, "failed to confirm relabel was added to secret"))
+		t.Fatal(errors.Wrap(err, "failed to create valid AlertRelabelConfig"))
 	}
 
+	// Check Prometheus config is taking the AlertRelabelConfig into account.
+	validateCurrentRelabelConfig(t, append(initialRelabelConfig, &relabel.Config{
+		SourceLabels: model.LabelNames{"alertname", "severity"},
+		Regex:        relabel.MustNewRegexp("Watchdog;none"),
+		TargetLabel:  "severity",
+		Replacement:  "critical",
+		Action:       "replace",
+		Separator:    ";",
+	}))
+
 	// Delete the AlertRelabelConfig.
-	err = relabelConfigs.Delete(ctx, arcName, metav1.DeleteOptions{})
+	err = relabelConfigs.Delete(ctx, arc.Name, metav1.DeleteOptions{})
 	if err != nil {
 		t.Fatal(errors.Wrap(err, "failed to delete AlertRelabelConfig"))
 	}
 
-	// Check that it is removed from the secret.
-	err = framework.Poll(time.Second, 2*time.Minute, func() error {
-		s, err := secrets.Get(ctx, relabelSecretName, metav1.GetOptions{})
-		if err != nil {
-			return err
-		}
-
-		var configsFromSecret []relabel.Config
-		err = yaml.Unmarshal([]byte(s.Data[relabelSecretKey]), &configsFromSecret)
-		if err != nil {
-			t.Fatal(errors.Wrap(err, "failed to unmarshal AlertRelabelConfig from secret"))
-		}
-
-		// Should only have the default config.
-		if len(configsFromSecret) != 1 {
-			return fmt.Errorf("Secret contains %d relabel configs, but only 1 expected",
-				len(configsFromSecret))
-		}
-
-		return nil
-	})
+	// Check Prometheus config forgot about the deleted AlertRelabelConfig.
+	validateCurrentRelabelConfig(t, initialRelabelConfig)
+}
 
+// prometheusRelabelConfig returns the alert relabel configuration part used by Prometheus config
+func prometheusRelabelConfig(t *testing.T) ([]*relabel.Config, error) {
+	prometheusConfig, err := f.PrometheusConfigFromSecret(t, f.Ns, prometheusConfigSecretName)
```

Comment:
If the function gets a t parameter, we can check the error directly?
```suggestion
	t.Helper()

	prometheusConfig, err := f.PrometheusConfigFromSecret(t, f.Ns, prometheusConfigSecretName)
	require.NoError(t, err)
```

Reply:
I thought about returning an error to be able to retry this as well if there are any intermittent problems, do you think we should be aware of those and fail from the first time?

Reply:
hmm right after reading again `PrometheusConfigFromSecret()` it doesn't seem to use to the `t` parameter so probably the best option is to remove it from the signature?

Reply:
Actually we pass it to `MustGetSecret` and we use it for `t.Helper()` now.

Reply:
Ah yes. Then maybe `PrometheusConfigFromSecret()` shouldn't return an error? If the secret data is invalid (e.g. can't be unzipped/deserialized), we should fail hard with `t.Fail`?

Reply:
That was my question :) I see that `MustGetSecret` already has retries that should deal with network/apiserver `intermittent problems`, we're left with unzipping/deserialization problems, which for sure we don't want to hide/we can never hide. Done.

Reply:
:+1: